server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -820,11 +878,11 @@ private boolean isIndexOderThan(
 
     /**
      * Checks if the provided backing index is managed by the data stream lifecycle as part of this data stream.
-     * If the index is not a backing index of this data stream, or we cannot supply its metadata
+     * If the index is not a backing index or a failure store index of this data stream, or we cannot supply its metadata
      * we return false.
      */
     public boolean isIndexManagedByDataStreamLifecycle(Index index, Function<String, IndexMetadata> indexMetadataSupplier) {
-        if (backingIndices.indices.contains(index) == false) {
+        if (backingIndices.containsIndex(index.getName()) == false && failureIndices.containsIndex(index.getName()) == false) {
```

Comment:
Was it an intentional change to use `containsIndex` here rather than the original `indices.contains`? The difference is that the former builds a lookup `Set`, whereas the latter just traverses the `List`. I guess it's a trade-off between performance and memory allocation. I'm inclined to say we don't generally need the performance win in this method. What do you think?

Reply:
It was intentional now that it exists at least. This method is used quite a bit with multiple indices and the same cluster state, so I thought it would be a good performance boost.

It's being used in DataStreamLifecycleService, when retrieving information or stats about a data stream, and in the explain of data stream lifecycle. I think all these cases could benefit from having the set calculated once and reused. But we could also separate this into a different PR if you are concerned about memory allocation.

Reply:
My reasoning was that since DSL runs in a background thread, performance is less of a priority (I'm not saying we should disregard it completely ofc). But I'm definitely fine with doing this in this PR, I just wanted to put it up for discussion. So no changes requested from me here :)

Reply:
Please, always raise concerns like this, through conversation we can uncover a lot of things.